lib/mlir/Dialect/Clift/Utils/CBackend.cpp

```diff
@@ -0,0 +1,1288 @@
+      if (It == C.Binary.Functions().end())
+        revng_abort("No matching model function.");
+      return &*It;
+    }
+
+    if (auto L = pipeline::locationFromString(revng::ranks::DynamicFunction,
+                                              Op.getUniqueHandle())) {
+      auto [Key] = L->at(revng::ranks::DynamicFunction);
+      auto It = C.Binary.ImportedDynamicFunctions().find(Key);
+      if (It == C.Binary.ImportedDynamicFunctions().end())
+        revng_abort("No matching model dynamic function.");
+      return &*It;
+    }
+
+    revng_abort("Unrecognizable function unique handle.");
+  }
+
+  const model::Function &getModelFunction(FunctionOp Op) {
+    auto Variant = getModelFunctionVariant(Op);
+    if (auto F = std::get_if<const model::Function *>(&Variant))
+      return **F;
+    revng_abort("Expected isolated model function.");
+  }
+
+  const model::TypeDefinition *
+  getModelTypeDefinition(uint64_t ID, model::TypeDefinitionKind::Values Kind) {
+    auto It = C.Binary.TypeDefinitions().find({ ID, Kind });
+    if (It == C.Binary.TypeDefinitions().end())
+      return nullptr;
+    return It->get();
+  }
+
+  const model::TypeDefinition &getModelTypeDefinition(TypeDefinitionAttr Type) {
+    using TypeKind = model::TypeDefinitionKind::Values;
+
+    if (mlir::isa<FunctionTypeAttr>(Type)) {
+      auto CF = getModelTypeDefinition(Type.id(),
+                                       TypeKind::CABIFunctionDefinition);
+      auto RF = getModelTypeDefinition(Type.id(),
+                                       TypeKind::RawFunctionDefinition);
+
+      if (CF != nullptr and RF != nullptr)
+        revng_abort("Ambiguous model function type definition");
+
+      if (CF != nullptr)
+        return *CF;
+
+      if (RF != nullptr)
+        return *RF;
+
+      revng_abort("No matching model function type definition.");
+    }
+
+    model::TypeDefinitionKind::Values Kind;
+    if (mlir::isa<TypedefTypeAttr>(Type))
+      Kind = model::TypeDefinitionKind::TypedefDefinition;
+    else if (mlir::isa<EnumTypeAttr>(Type))
+      Kind = model::TypeDefinitionKind::EnumDefinition;
+    else if (mlir::isa<StructTypeAttr>(Type))
+      Kind = model::TypeDefinitionKind::StructDefinition;
+    else if (mlir::isa<UnionTypeAttr>(Type))
+      Kind = model::TypeDefinitionKind::UnionDefinition;
+    else
+      revng_abort("Unsupported type definition");
+
+    auto ModelType = getModelTypeDefinition(Type.id(), Kind);
+    if (ModelType == nullptr)
+      revng_abort("No matching model type definition.");
+    return *ModelType;
+  }
+
+  void emitPrimitiveType(PrimitiveType Type) {
+    auto Kind = static_cast<model::PrimitiveKind::Values>(Type.getKind());
+    auto ModelType = model::PrimitiveType::make(Kind, Type.getSize());
+    Out << C.getLocationReference(llvm::cast<model::PrimitiveType>(*ModelType));
+  }
+
+  RecursiveCoroutine<void>
+  emitDeclaration(ValueType Type,
+                  std::optional<llvm::StringRef> DeclaratorName) {
```

Comment:
Any specific reason why you're re-implementing this? Can't you translate `Type` to a `model::Type` and then use `getNamedCInstance`?

`getNamedCInstance` also takes care of type inlining, which is disabled as of now, but using that would make it effortless to enable it also for the new backend.

Reply:
If nothing else, it seems very inefficient to construct a model type when we emitting the Clift type directly is trivial.

> `getNamedCInstance` also takes care of type inlining

Does it? I couldn't see anything related to type inlining in its implementation. Maybe I missed it.